bitcoin/src/pow.rs

```diff
@@ -366,6 +366,56 @@ impl CompactTarget {
         Ok(Self::from_consensus(lock_time))
     }
 
+    /// Computes the [`CompactTarget`] from a difficulty adjustment.
+    ///
+    /// ref: <https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp>
+    ///
+    /// Given the previous Target, represented as a [`CompactTarget`], the difficulty is adjusted
+    /// by taking the timespan between them, and multipling the current [`CompactTarget`] by a factor
+    /// of the net timespan and expected timespan. The [`CompactTarget`] may not adjust by more than
+    /// a factor of 4, or adjust beyond the maximum threshold for the network.
+    ///
+    /// # Note
+    ///
+    /// Under the consensus rules, the different in the number of blocks between the headers does
+    /// not equate to the `difficulty_adjustment_interval` of [`Params`]. This is due to an off-by-one
+    /// error, and, the expected number of blocks in between headers is `difficulty_adjustment_interval - 1`
+    /// when calculating the difficulty adjustment.
+    ///
+    /// Take the example of the first difficulty adjustment. Block 2016 introduces a new [`CompactTarget`],
+    /// which takes the net timespan between Block 2015 and Block 0, and recomputes the difficulty.
+    ///
+    /// # Returns
+    ///
+    /// The expected [`CompactTarget`] recalculation.
+    pub fn from_next_work_required(
+        last: CompactTarget,
+        timespan: u64,
+        params: impl AsRef<Params>,
+    ) -> CompactTarget {
+        let params = params.as_ref();
+        if params.no_pow_retargeting {
+            return last;
+        }
+        let mut actual_timespan = timespan;
+        let prev_target: Target = last.into();
+        if timespan.le(&(params.pow_target_timespan / 4)) {
+            actual_timespan = params.pow_target_timespan / 4;
+        }
+        if timespan.ge(&(params.pow_target_timespan * 4)) {
+            actual_timespan = params.pow_target_timespan * 4;
+        }
```

Comment:
All this can be written with the very nice `clamp` function (I learned that from @Kixunil when I was working on #2180 :)

```rust
        let min_timespan = params.pow_target_timespan >> 2;
        let max_timespan = params.pow_target_timespan << 2;
        let actual_timespan = timespan.clamp(min_timespan, max_timespan);
```

Reply:
ref: https://doc.rust-lang.org/std/cmp/trait.Ord.html#method.clamp

Reply:
Ah, beautiful. I think this makes the code more readable. I don't see a problem adding it.